athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchRecordHandler.java

```diff
@@ -0,0 +1,215 @@
+ */
+public class ElasticsearchRecordHandler
+        extends RecordHandler
+{
+    private static final Logger logger = LoggerFactory.getLogger(ElasticsearchRecordHandler.class);
+
+    /**
+     * used to aid in debugging. Athena will use this name in conjunction with your catalog id
+     * to correlate relevant query errors.
+     */
+    private static final String SOURCE_TYPE = "elasticsearch";
+
+    // Pagination batch size (100 documents).
+    private static final int QUERY_BATCH_SIZE = 100;
+
+    private AmazonS3 amazonS3;
+
+    private final AwsRestHighLevelClientFactory clientFactory;
+
+    public ElasticsearchRecordHandler()
+    {
+        this(AmazonS3ClientBuilder.defaultClient(), AWSSecretsManagerClientBuilder.defaultClient(),
+                AmazonAthenaClientBuilder.defaultClient(), ElasticsearchHelper.getClientFactory());
+    }
+
+    @VisibleForTesting
+    protected ElasticsearchRecordHandler(AmazonS3 amazonS3, AWSSecretsManager secretsManager, AmazonAthena amazonAthena,
+                                         AwsRestHighLevelClientFactory clientFactory)
+    {
+        super(amazonS3, secretsManager, amazonAthena, SOURCE_TYPE);
+        this.amazonS3 = amazonS3;
+        this.clientFactory = clientFactory;
+    }
+
+    /**
+     * Used to read the row data associated with the provided Split.
+     *
+     * @param spiller A BlockSpiller that should be used to write the row data associated with this Split.
+     * The BlockSpiller automatically handles chunking the response, encrypting, and spilling to S3.
+     * @param recordsRequest Details of the read request, including:
+     * 1. The Split
+     * 2. The Catalog, Database, and Table the read request is for.
+     * 3. The filtering predicate (if any)
+     * 4. The columns required for projection.
+     * @param queryStatusChecker A QueryStatusChecker that you can use to stop doing work for a query that has already terminated
+     * @throws IOException
+     * @note Avoid writing >10 rows per-call to BlockSpiller.writeRow(...) because this will limit the BlockSpiller's
+     * ability to control Block size. The resulting increase in Block size may cause failures and reduced performance.
+     */
+    @Override
+    protected void readWithConstraint(BlockSpiller spiller, ReadRecordsRequest recordsRequest,
+                                      QueryStatusChecker queryStatusChecker)
+    {
+        logger.info("readWithConstraint - enter\n\nDomain: {}\n\nIndex: {}\n\nMapping: {}\n\nConstraints: {}",
+                recordsRequest.getTableName().getSchemaName(), recordsRequest.getTableName().getTableName(),
+                recordsRequest.getSchema(), recordsRequest.getConstraints().getSummary());
+
+        String endpoint = ElasticsearchHelper.getDomainEndpoint(recordsRequest.getTableName().getSchemaName());
+        long numRows = 0;
+        AtomicLong numResultRows = new AtomicLong(0);
```

Comment:
Why atomic long? this runs in 1 thread

Reply:
Code was copied from docdb connector. Removed numResultRows - not necessary.

Reply:
Ah ok, its because it is used in the following Lambda and the IDE will complain that it needs to be effectively final.